### DISCUSSION TITLE
Derivative of GP w.r.t. Input

### DISCUSSION THREAD

**Discussion by ejmeitz** [GPflow/GPflow#2125] [Q&A]
I'm looking around at the various GP packages and trying to find one that supports taking the derivative with respect to the input of the GP. I looked through the docs here and did not see an example or mention of this but since GPFlow is implemented in TensorFlow I figured automatic differentiation might allow this to work easily.

I'm pretty sure the derivative of a GP is just another GP so it would be super nice if there was a way with automatic differentiation just get another GP object back, but I would also settle for a manual implementation of a point evaluation as long as I can get a back a value with uncertainty. Any help figuring this out would be appreciated, thanks!

**Comment by st--** [comment]
If all you care about is the gradient of the mean function, then you can just autodiff through `model.predict_f`. But the gradient of the predictive variance is not the same as the variance of the predicted gradient: to get the uncertainty right, in the GP prediction equations you have to replace the $K_{fy}$ from the original $K_{fy} = k(x^*, x_D)$ to the gradient prediction $K'_{f'y} = \frac{\partial}{\partial x^*} k(x^*, x_D)$. Of course this gradient you could compute by autodiff (though it's just as possible in closed form), but you'd still have to change the code so that it uses a different kernel for some of the kernel matrices... hope this gives you a starting point to develop it!
